### PR TITLE
 Initial implementation of static analysis listeners

### LINKED ISSUE #280: Findings listed by Teamscale
The recent bunch of changes has introduced a couple of findings in Teamscale, most notably many public methods are left undocumented, which does not need to be changed. The remaining issues are not hard to address.

Steps to fix:
1. Create account for Teamscale on http://demo.teamscale.com
2. Inspect Findings view
3. Sort by date
4. Filter out documentation-related warnings
5. Fix findings introduced in 2018

screenshot from 2018-02-07 09-23-20

### PR DESCRIPTION
**Pull request by TimvdLippe**
Built on top of #278, interesting commit is https://github.com/TestRoots/watchdog/commit/6e723a2d5c453b768bb317c9933a7208130e0a6a.

This PR introduces an initial implementation of the static analysis listener. The hardest and most confusing part of the listener is to start listening at the correct time. In IntelliJ, there is no way to know which warnings were generated based on the "initial pass" of the code analyzer when you open your editor. Since these warnings are already tracked, we want to ignore these, but do listen to subsequent changes.

The logic to make sure we only listen for the second batch and further is as follows:
1. First we have to wait till the editor finished indexing. This makes sure that the initial code analyzer "success" event is dodged.
2. Then once we receive the global code analyzer event, we need to check whether this event is actually for the file that we were tracking. To do so, we need to check the `FileDirtyStatus` of the document we are watching.

If both events happened, then we start listening for warnings that are added and removed from the model. If they are a `WARNING`, the trackingManager receives the corresponding event. This is `UNKNOWN` for now, as I have not anonymized the warnings yet.

Note that this sends ALL warnings to the server. We will probably want to do some simple clustering (for a timeline) on the client first, to not overload the server with events. I did introduced a concept of delayed processing (`warning.creationTime.plus(CREATION_TRIGGERED_DELAY).isBeforeNow()`), to cluster at that point in time.

We might want to empirically determine the value of `CREATION_TRIGGERED_DELAY` to see how granular we want to make our data. This is maybe something to discuss with @azaidman ?

All in all, I would want to hold off publishing to WatchDog, before we implement the clustering logic, to prevent overloading the server at this point.

Fixes #280

### CODE REVIEW

**Comment by TimvdLippe** [comment]
@Inventitech All right, this should be the initial implementation of the listeners in both Eclipse and IntelliJ. In Intellij we had to do some dirty work to get the correct event. In Eclipse, the diffing algorithm was necessary as Eclipse trashes and recreates all markers.

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            });
+        });
+
+        return markupModelListener;
+    }
+
+    @Override
+    public void afterAdded(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        if (rangeHighlighterEx.getLayer() == HighlighterLayer.WARNING) {
+            addCreatedWarning(StaticAnalysisType.UNKNOWN);
```
Comment (TimvdLippe) [inline intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java]: Instead of flushing every single warning, we could use the Eclipse model of processing all warnings on save. It might be possible with http://arhipov.blogspot.nl/2011/04/code-snippet-intercepting-on-save.html to call the `add{Created,Removed}Warning()`. Then whenever warnings are removed, we have to clear up our internal storage of these two files and then "flush" when the save action occurs. WDYT @Inventitech ?

**Comment by Inventitech** [comment]
We are going to need a couple of tests to verify the base functionality of this. :)

**Comment by TimvdLippe** [comment]
@Inventitech Current status of this PR: It took a lot of debugging, but I finally got the Eclipse UI tests passing on Travis. This means that the test story for Eclipse is solid now. I still need to update the IntelliJ implementation to only work on `save`.

**Comment by Inventitech** [comment]
I know this is WIP (and great W so far!), but could you the tests seem to have grown pretty lengthy and complicated to understand. Could you describe them in method comments? Thanks!

**Comment by TimvdLippe** [comment]
@Inventitech In general I prefer lengthy tests that clearly describe (in code) each test, rather than shortened tests. To me, the tests are very clear. Could you therefore add some comments on the test lines that you are having trouble understanding, then I can add comments there? :smile:

File: intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java (hunk 1 above)
Comment (Inventitech) [reply intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java]: We discussed this in a meeting

**Comment by TimvdLippe** [comment]
@Inventitech Implementation is mostly finished now. I am going to try to write some IntelliJ tests to make sure that is working as well. I think it would be best if we try this out locally first before publishing a new version.

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		private int[][] memoization;
+		private int oldSize;
+		private int currentSize;
+
+		MarkerBackTrackingAlgorithm(List<MarkerHolder> oldMarkers, List<MarkerHolder> currentMarkers) {
+			this.oldMarkers = oldMarkers;
+			this.currentMarkers = currentMarkers;
+			this.oldSize = oldMarkers.size();
+			this.currentSize = currentMarkers.size();
+			memoization = new int[oldSize + 1][currentSize + 1];
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: this.
Relevant: https://www.youtube.com/watch?v=M5d7vygUPoQ

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	 * of cost of modifications to the lists. It then traverses through this matrix
+	 * and based on the cost either triggers a deletion or an addition.
+	 *
+	 * There are multiple viable implementations of this diffing
+	 * algorithm and this abstraction allows us to later change the actual
+	 * implementation without breaking the usage contract.
+	 *
+	 * @see https://en.wikipedia.org/wiki/Longest_common_subsequence_problem
+	 */
+	class MarkerBackTrackingAlgorithm {
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: visibility?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
+package nl.tudelft.watchdog.eclipse.logic.ui.listeners;
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: Can we break this apart into its class constiuents and move them to their own package?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	@Override
+	public void resourceChanged(IResourceChangeEvent event) {
+		try {
+			this.oldFileMarkers = this.currentFileMarkers;
+			this.currentFileMarkers = new HashMap<>();
+			event.getDelta().accept(this.createVisitor());
+		} catch (CoreException e) {
+			// TODO Auto-generated catch block
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: Please address all TODO comments

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/GeneralActivityListener.java [hunk 6]
```diff
@@ -15,13 +15,15 @@
...
 	/** Constructor. */
 	public GeneralActivityListener(Display display) {
-		display.addFilter(SWT.KeyDown | SWT.KeyUp | SWT.MouseDown | SWT.MouseUp
-				| SWT.MouseMove, new Listener() {
-
-			@Override
-			public void handleEvent(Event event) {
-				WatchDogEventType.USER_ACTIVITY.process(event);
-			}
+		display.syncExec(() -> {
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/GeneralActivityListener.java]: Why this change?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/WorkbenchListener.java [hunk 7]
```diff
@@ -21,7 +25,7 @@
 	private TransferManager transferManager;
 
 	/** The debug event manager used to process debug events. */
-	private TrackingEventManager TrackingEventManager;
+	private TrackingEventManager trackingEventManager;
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/WorkbenchListener.java]: good catch

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/ui/handlers/StartupHandler.java [hunk 8]
```diff
@@ -44,6 +44,7 @@ public static void startWatchDog() {
 			UIUtils.refreshCommand(UIUtils.COMMAND_SHOW_INFO);
 			updateView();
 		} catch (Exception exception) {
+			exception.printStackTrace();
```
Comment (Inventitech) [inline eclipse/plugin/src/nl/tudelft/watchdog/eclipse/ui/handlers/StartupHandler.java]: We don't want stdout cluttered. Wouldn't this be better in the log?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java (hunk 2 above)
Comment (TimvdLippe) [reply eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: I think you are reviewing an old commit. This is fixed in the overall PR

File: eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	public void tearDown() throws Exception {
+		this.project.delete(true, true, null);
+		this.workspace.save(true, null);
+
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
+		this.workbenchListener.shutDown();
+	}
+
+	@Test
+	public void picks_up_created_warnings_on_first_save() throws Exception {
```
Comment (Inventitech) [inline eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java]: These method names do not exactly follow camel casing

File: eclipse/tests/src/nl/tudelft/watchdog/logic/ui/WatchDogEventManagerTest.java [hunk 10]
```diff
@@ -39,10 +40,13 @@
  * eventually transfered to the server, this is one of the most crucial parts of
  * WatchDog. Tests could flicker because they deal with timers (and Java gives
  * no guarantee as to when these timers will be executed).
+ *
+ * TODO(timvdlippe): This test is disabled until we can properly deflake it.
```
Comment (Inventitech) [inline eclipse/tests/src/nl/tudelft/watchdog/logic/ui/WatchDogEventManagerTest.java]: Move this to right above the @Ignore?

**Comment by Inventitech** [comment]
Current tests contain at least a class documentation. Especially for tests longer than 10 lines, I want a brief summary to understand what it does. Most tests are not difficult, but the amount of code takes a lot of time to read. Sometimes, I am only interested in what scenarios have been covered. I can't get this info right now without spending a good 10 minutes.

File: eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+	@Test
+	public void does_not_generate_events_for_same_markers() throws Exception {
+		IMarker marker = this.testFile.createMarker(IMarker.PROBLEM);
+		marker.setAttribute(IMarker.LINE_NUMBER, 1);
+		IMarker marker2 = this.testFile.createMarker(IMarker.PROBLEM);
+		marker2.setAttribute(IMarker.LINE_NUMBER, 2);
+
+		this.workspace.save(true, null);
+		Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
```
Comment (Inventitech) [inline eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java]: This line is hard to understand.

File: eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	@Test
+	public void generates_removal_after_marker_is_deleted() throws Exception {
+		IMarker marker = this.testFile.createMarker(IMarker.PROBLEM);
+		marker.setAttribute(IMarker.LINE_NUMBER, 1);
+		IMarker marker2 = this.testFile.createMarker(IMarker.PROBLEM);
+		marker2.setAttribute(IMarker.LINE_NUMBER, 2);
+
+		this.workspace.save(true, null);
```
Comment (Inventitech) [inline eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java]: What is the general rule for this. in the rest of the project? I don't think we usually use it outside constructors, right?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java (hunk 4 above)
Comment (TimvdLippe) [reply eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: They are part of the whole implementation and not intended to be used outside of this class. Moreover, they reference methods/fields in the outer class, so this would make the logic more complicated.

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/GeneralActivityListener.java (hunk 6 above)
Comment (TimvdLippe) [reply eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/GeneralActivityListener.java]: There is no guarantee that this listener runs on the UI Thread. This actually happens when you run in headless mode in the tests. Therefore, the method has to be invoked on that UI Thread with `display.syncExec`

File: eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java (hunk 9 above)
Comment (TimvdLippe) [reply eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java]: This is a personal preference. We use this casing in Mockito. You want to change it?

File: eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java (hunk 12 above)
Comment (TimvdLippe) [reply eclipse/tests/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/MarkupModelListenerTest.java]: I am not sure what you mean. What do you mean with "outside constructors"?

File: eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java (hunk 3 above)
Comment (TimvdLippe) [reply eclipse/plugin/src/nl/tudelft/watchdog/eclipse/logic/ui/listeners/EclipseMarkupModelListener.java]: This class is intended to be an implementation detail of the listener and thus should not be `public`. Could be `private`, but that breaks the Javadoc `{@link}` in Eclipse :cry:

File: eclipse/tests/src/nl/tudelft/watchdog/logic/ui/WatchDogEventManagerTest.java (hunk 10 above)
Comment (TimvdLippe) [reply eclipse/tests/src/nl/tudelft/watchdog/logic/ui/WatchDogEventManagerTest.java]: That would break the JavaDoc generation. I have moved the `@Ignore` to the first line

**Comment by TimvdLippe** [comment]
@Inventitech Most of the feedback has been processed.

**Comment by TimvdLippe** [comment]
@Inventitech Friendly ping for next week :smile:

**Comment by TimvdLippe** [comment]
Note that this effort was split into two. This PR contains the initial implementation. The code is later revised and extended in #282 which includes the warning classification. The build fully works if you first merge this branch and then #282.

**Comment by TimvdLippe** [comment]
@Inventitech In the last commit I have fixed the updatesite. Apparently the Eclipse plugin has been published ever since, but users never got an update as the updatesite url was wrong (it was not pointing to an actual p2 repository).

To resolve this issue, I propose to turn http://updatesite.testroots.org/ into an p2 repository itself (by simply copying from `/updatesite/` to `/`). This should start notifying existing users of the actual update. Then with this fix, they will actually be redirected to the correct location and then should continue receiving updates.
